src/Microsoft.ML.PCA/PcaTrainer.cs

```diff
@@ -91,7 +91,7 @@ public class Arguments : UnsupervisedLearnerInputBaseWithWeight
         /// <param name="oversampling">Oversampling parameter for randomized PCA training.</param>
         /// <param name="center">If enabled, data is centered to be zero mean.</param>
         /// <param name="seed">The seed for random number generation.</param>
-        public RandomizedPcaTrainer(IHostEnvironment env,
+        internal RandomizedPcaTrainer(IHostEnvironment env,
```

Comment:
Shouldn't we just make the class `Internal`/`BestFriend` and keep this `public`?

Reply:
not really. 

we would want to expose these through mlcontext. not via *constructors*

---
In reply to: [254391018](https://github.com/dotnet/machinelearning/pull/2390#discussion_r254391018) [](ancestors = 254391018)

Reply:
Resolved. I hadn't seen the pattern for trainable transforms where the class is `public` and methods are `internal`.

Reply:
in ML.NET  terms,  this is "trainer estimator"  for anomaly detection tasks.  most other  "trainer estimator"s follow the same pattern  e.g.  KMeansPlusPlusTrainer

---
In reply to: [255587702](https://github.com/dotnet/machinelearning/pull/2390#discussion_r255587702) [](ancestors = 255587702)